core/node/track_streams/multi_sync_runner.go

```diff
@@ -0,0 +1,776 @@
+		otelTracer:             otelTracer,
+	}
+}
+
+// Run starts the operation of the MultiSyncRunner and continues to add streams to sync sessions until rootCtx is canceled.
+func (msr *MultiSyncRunner) Run(
+	rootCtx context.Context,
+) {
+	for range msr.config.NumWorkers {
+		// Each worker routine will continuously loop, placing streams in sync sessions, until the rootCtx is canceled.
+		msr.workerPool.Submit(func() {
+			for {
+				select {
+				case <-rootCtx.Done():
+					return
+				case streamInitRecord := <-msr.streamsToSync:
+					msr.addToSync(rootCtx, streamInitRecord)
+				}
+			}
+		})
+	}
+
+	<-rootCtx.Done()
+	msr.workerPool.StopWait()
+}
+
+func (msr *MultiSyncRunner) getNodeRegistry() nodes.NodeRegistry {
+	idx := rand.Int63n(int64(len(msr.nodeRegistries)))
+	return msr.nodeRegistries[idx]
+}
+
+// addToSync adds a stream to a sync session for the stream's current sticky peer. We advance through the
+// node's peers if any attempt to add the stream to a sync session fails.
+func (msr *MultiSyncRunner) addToSync(
+	rootCtx context.Context,
+	record streamSyncInitRecord,
+) {
+	targetNode := record.remotes.GetStickyPeer()
+	pool := msr.getNodeRequestPool(targetNode)
+	log := logging.FromCtx(rootCtx)
+
+	sessionRunner, ok := msr.unfilledSyncs.Load(targetNode)
+	var runner *syncSessionRunner
+	if !ok {
+		runner = NewSyncSessionRunner(
+			rootCtx,
+			msr.streamsToSync,
+			msr.getNodeRegistry(),
+			func(streamId shared.StreamId, streamAndCookie *protocol.StreamAndCookie) (events.TrackedStreamView, error) {
+				return msr.trackedViewConstructor(rootCtx, streamId, msr.onChainConfig, streamAndCookie)
+			},
+			msr.config.StreamsPerSyncSession,
+			targetNode,
+			msr.metrics,
+			msr.otelTracer,
+		)
+		var loaded bool
+
+		// Running the runner involves a request to the node to establish the sync. We use the
+		// workerpool for the node to rate limit these.
+
+		// Pre-emptively acquire a connection for the newly created runner above so that if the
+		// store is successful, we are guaranteed to call it's 'Run' method. This will keep
+		// other workers from becoming indefinitely blocked on stream insertion to this runner.
+		if err := pool.Acquire(rootCtx, 1); err != nil {
+			if !errors.Is(err, context.Canceled) {
+				log.Errorw(
+					"unable to acquire worker pool task for node; closing runner and re-assigning stream",
+					"error",
+					err,
+					"node",
+					targetNode,
+					"streamId",
+					record.streamId,
+				)
+				msr.streamsToSync <- record
+			}
+			runner.Close(fmt.Errorf("unable to acquire worker pool task for node: %w", err))
+			return
+		}
+
+		if sessionRunner, loaded = msr.unfilledSyncs.LoadOrStore(targetNode, runner); !loaded {
+			// If our new runner won the race to be stored for this node, kick off the runner. Streams
+			// are not assignable until the sync session starts.
+			go runner.Run(rootCtx)
+			runner.WaitUntilStarted()
```

Comment:
Scenario:
Runner doesn't exists, 2 go-routines call at the same time `func (msr *MultiSyncRunner) addToSync` for a stream that ends up with the same target node. One of them wins the race and waits for the runner until it is started. The other one looses and continues immediately. It can call AddStream on a runner that isn't fully started which can fail and can try to initialises a new runner that replaces the one that is currently starting up.

Apart from releasing a semaphore nothing happens. Maybe it is possible to move the `runner.WaitUntilStarted()` a couple of lines down after `runner = sessionRunner.(*syncSessionRunner)`?